src/MongoDB/Session.c

```diff
@@ -228,6 +230,153 @@ static PHP_METHOD(Session, getOperationTime)
+	}
+
+	if (php_array_existsc(options, "readPreference")) {
+		zval* read_preference = php_array_fetchc(options, "readPreference");
+
+		if (Z_TYPE_P(read_preference) != IS_OBJECT || !instanceof_function(Z_OBJCE_P(read_preference), php_phongo_readpreference_ce TSRMLS_CC)) {
+			phongo_throw_exception(PHONGO_ERROR_INVALID_ARGUMENT TSRMLS_CC, "Expected \"readPreference\" option to be %s, %s given", ZSTR_VAL(php_phongo_readpreference_ce->name), PHONGO_ZVAL_CLASS_OR_TYPE_NAME_P(read_preference));
+			return false;
+		}
+
+		if (!opts) {
+			opts = mongoc_transaction_opts_new();
+		}
+
+		mongoc_transaction_opts_set_read_prefs(opts, phongo_read_preference_from_zval(read_preference TSRMLS_CC));
+	}
+
+	if (php_array_existsc(options, "writeConcern")) {
+		zval* write_concern = php_array_fetchc(options, "writeConcern");
+
+		if (Z_TYPE_P(write_concern) != IS_OBJECT || !instanceof_function(Z_OBJCE_P(write_concern), php_phongo_writeconcern_ce TSRMLS_CC)) {
+			phongo_throw_exception(PHONGO_ERROR_INVALID_ARGUMENT TSRMLS_CC, "Expected \"writeConcern\" option to be %s, %s given", ZSTR_VAL(php_phongo_writeconcern_ce->name), PHONGO_ZVAL_CLASS_OR_TYPE_NAME_P(write_concern));
+			return false;
+		}
+
+		if (!opts) {
+			opts = mongoc_transaction_opts_new();
+		}
+
+		mongoc_transaction_opts_set_write_concern(opts, phongo_write_concern_from_zval(write_concern TSRMLS_CC));
+	}
+
+	return opts;
+}
+
+/* {{{ proto void MongoDB\Driver\Session::startTransaction([array $options = null])
+   Starts a new transaction */
+static PHP_METHOD(Session, startTransaction)
+{
+	php_phongo_session_t*     intern;
+	zval*                     options     = NULL;
+	mongoc_transaction_opt_t* txn_options = NULL;
+	bson_error_t              error;
+	SUPPRESS_UNUSED_WARNING(return_value_ptr)
+	SUPPRESS_UNUSED_WARNING(return_value_used)
+
+	intern = Z_SESSION_OBJ_P(getThis());
+
+	if (zend_parse_parameters(ZEND_NUM_ARGS() TSRMLS_CC, "|a", &options) == FAILURE) {
+		return;
+	}
+
+	if (options) {
+		txn_options = php_mongodb_session_parse_transaction_options(options TSRMLS_CC);
+	}
+	if (EG(exception)) {
+		return;
+	}
+
+	if (!mongoc_client_session_start_transaction(intern->client_session, txn_options, &error)) {
+		phongo_throw_exception_from_bson_error_t(&error TSRMLS_CC);
+	}
+
+	if (txn_options) {
+		mongoc_transaction_opts_destroy(txn_options);
+	}
+} /* }}} */
+
+/* {{{ proto void MongoDB\Driver\Session::commitTransaction(void)
+   Commits an existing transaction */
+static PHP_METHOD(Session, commitTransaction)
+{
+	php_phongo_session_t* intern;
+	bson_error_t          error;
+	bson_t                reply;
+	SUPPRESS_UNUSED_WARNING(return_value_ptr)
+	SUPPRESS_UNUSED_WARNING(return_value_used)
+
+	intern = Z_SESSION_OBJ_P(getThis());
+
+	if (zend_parse_parameters_none() == FAILURE) {
+		return;
+	}
+
+	if (!mongoc_client_session_commit_transaction(intern->client_session, &reply, &error)) {
+		phongo_throw_exception_from_bson_error_and_reply_t(&error, &reply TSRMLS_CC);
```

Comment:
Test this for leaks. I think you still need a `bson_destroy(&reply);` after this.

`executeCommand()` does so before returning as its `free_reply` boolean will be true when it calls `phongo_throw_exception_from_bson_error_and_reply_t`.